src/roblox_api/open_cloud.rs

```diff
@@ -10,83 +13,79 @@ use rbxcloud::rbx::{
 
-use super::{
-    legacy::LegacyClient, ImageUploadData, RobloxApiClient, RobloxApiError, RobloxCredentials,
-    UploadResponse,
-};
+use super::{ImageUploadData, RobloxApiClient, RobloxApiError, RobloxCredentials, UploadResponse};
 
-pub struct OpenCloudClient {
+pub struct OpenCloudClient<'a> {
     credentials: RobloxCredentials,
     creator: AssetCreator,
     assets: RbxAssets,
-    runtime: Runtime,
+    _marker: PhantomData<&'a ()>,
 }
 
-impl RobloxApiClient for OpenCloudClient {
-    fn new(credentials: RobloxCredentials) -> Result<Self, RobloxApiError> {
+#[async_trait]
+impl<'a> RobloxApiClient<'a> for OpenCloudClient<'a> {
+    fn new(credentials: RobloxCredentials) -> Result<Self> {
         let creator = match (credentials.group_id, credentials.user_id) {
-            (Some(id), None) => Ok(AssetCreator::Group(AssetGroupCreator {
+            (Some(id), None) => AssetCreator::Group(AssetGroupCreator {
                 group_id: id.to_string(),
-            })),
-            (None, Some(id)) => Ok(AssetCreator::User(AssetUserCreator {
+            }),
+            (None, Some(id)) => AssetCreator::User(AssetUserCreator {
                 user_id: id.to_string(),
-            })),
-            (None, None) => Err(RobloxApiError::ApiKeyNeedsCreatorId),
-            (Some(_), Some(_)) => Err(RobloxApiError::AmbiguousCreatorType),
-        }?;
-
-        let assets = RbxCloud::new(
-            credentials
-                .api_key
-                .as_ref()
-                .ok_or(RobloxApiError::MissingAuth)?
-                .expose_secret(),
-        )
-        .assets();
+            }),
+            _ => unreachable!(),
+        };
+
+        let Some(api_key) = credentials.api_key.as_ref() else {
+            bail!(RobloxApiError::MissingAuth);
+        };
+
+        let assets = RbxCloud::new(api_key.expose_secret()).assets();
 
         Ok(Self {
             creator,
             assets,
             credentials,
-            runtime: Runtime::new().unwrap(),
+            _marker: PhantomData::default(),
         })
     }
 
-    fn upload_image_with_moderation_retry(
-        &mut self,
-        data: &ImageUploadData,
-    ) -> Result<UploadResponse, RobloxApiError> {
-        match self.upload_image(data) {
-            Err(RobloxApiError::ResponseError { status, body })
-                if status == 400 && body.contains("moderated") =>
-            {
-                log::warn!(
-                    "Image name '{}' was moderated, retrying with different name...",
-                    data.name
-                );
-                self.upload_image(&ImageUploadData {
-                    name: "image",
-                    ..data.to_owned()
-                })
-            }
-
-            result => result,
-        }
-    }
-
-    fn upload_image(&mut self, data: &ImageUploadData) -> Result<UploadResponse, RobloxApiError> {
-        self.upload_image_inner(data)
+    // this was a bad idea, sorry
```

Comment:
Should we remove this?

Reply:
We need that method? Because the image will not upload if it is moderated.

Reply:
>:(

Reply:
\>:(